src/csrc/include/internal/rl/rollout_buffer.h

```diff
@@ -0,0 +1,314 @@
+ * SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER
+ * CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY,
+ * OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+ * OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+ */
+
+#pragma once
+#include <deque>
+#include <random>
+#include <tuple>
+
+#include <cuda_runtime.h>
+
+#include <c10/cuda/CUDAGuard.h>
+#include <c10/cuda/CUDAStream.h>
+#include <torch/torch.h>
+
+#include "internal/defines.h"
+#include "internal/rl/rl.h"
+
+namespace torchfort {
+
+namespace rl {
+
+using BufferEntry = std::tuple<torch::Tensor, torch::Tensor, float, float, float, bool>;
+
+enum RewardReductionMode { Sum = 1, Mean = 2, WeightedMean = 3, SumNoSkip = 4, MeanNoSkip = 5, WeightedMeanNoSkip = 6 };
+
+// abstract base class for rollout buffer
+class RolloutBuffer {
+public:
+  // disable copy constructor
+  RolloutBuffer(const RolloutBuffer&) = delete;
+  // base constructor
+  RolloutBuffer(size_t size, int device) : size_(size), device_(get_device(device)), last_episode_starts_(true) {}
+
+  // virtual functions
+  virtual void update(torch::Tensor, torch::Tensor, float, float, float, bool) = 0;
+  virtual void finalize(float, bool) = 0;
+  virtual std::tuple<torch::Tensor, torch::Tensor, torch::Tensor, torch::Tensor, torch::Tensor, torch::Tensor>
+  sample(int) = 0;
+  virtual bool isReady() const = 0;
+  virtual void reset(bool start_new_episode) = 0;
+  virtual void printInfo() const = 0;
+  virtual void save(const std::string& fname) const = 0;
+  virtual void load(const std::string& fname) = 0;
+  virtual torch::Device device() const = 0;
+
+protected:
+  size_t size_;
+  torch::Device device_;
+  bool last_episode_starts_;
+};
+
+class GAELambdaRolloutBuffer : public RolloutBuffer, public std::enable_shared_from_this<RolloutBuffer> {
+
+public:
+  // constructor
+  GAELambdaRolloutBuffer(size_t size, float gamma, float lambda, int device)
+    : RolloutBuffer(size, device), finalized_(false), gamma_(gamma), lambda_(lambda), rng_(), returns_(size), advantages_(size) {}
+
+  // disable copy constructor
+  GAELambdaRolloutBuffer(const GAELambdaRolloutBuffer&) = delete;
+
+  // update
+  void update(torch::Tensor s, torch::Tensor a, float r, float q, float log_p, bool d) {
+
+    // add no grad guard
+    torch::NoGradGuard no_grad;
+    
+    // do not push anything if buffer is full, instead check if buffer is finalized
+    if (buffer_.size() == size_) {
+      if (!finalized_) {
+	finalize(q, d);
+      }
+    } else {
+    
+      // clone the tensors and move to device
+      auto sc = s.to(device_, s.dtype(), /* non_blocking = */ false, /* copy = */ true);
+      auto ac = a.to(device_, a.dtype(), /* non_blocking = */ false, /* copy = */ true);
+
+      // 
+      
+      // add the newest/latest element in back
+      // TODO: is that right? should be
```

Comment:
Remove TODO if you've resolved the question :)

Reply:
fixed